testing/src/context.py

```diff
@@ -674,7 +674,7 @@ def run(self, event: "_Event", state: "State") -> "State":
             if self._action_failure_message is not None:
                 raise ActionFailed(
                     self._action_failure_message,
-                    state=self._output_state,
+                    state=self._output_state,  # type: ignore
```

Comment:
I guess it isn't worth the pain of bringing in `ops._private.harness.ActionOutput` here and `cast`ing or trying to use it instead of the `"State"` annotation for `_output_state`?

Reply:
It is bringing in `ops._private.harness.ActionOutput`. pyright was complaining about `state` being a `State` rather than `State|None` (there's an assert just above this that forces that), which I didn't really understand since it seems like it should be fine to be *more* specific. I basically punted on this one, figuring that there needed to be a bunch of "decease type:ignore usage" work later anyway and it could be figured out then, since it doesn't really gain much to have this be checked.

Reply:
Because of the following lines, and the fact that you are already asserting that it's not `None`, I assumed that the type checker was complaining about it being a `State` rather than a `ActionOutput`, which is the type that `ActionFailed` expects, but I haven't looked deeper, so perhaps I'm wrong. And probably not worth the pain either way.

```
    from scenario.state import (
        ...
        State,
        ...
    )
```

```python
        self._output_state: Optional["State"] = None
```

Reply:
`ActionFailed` is expecting a `State` (it expects an `ActionOutput|None` for the `output` argument, which defaults to `None`).

If I don't have the `type: ignore` then this is what I get:

```
/tmp/operator/testing/src/context.py
  /tmp/operator/testing/src/context.py:678:27 - error: Argument of type "State" cannot be assigned to parameter "state" of type "State | None" in function "__init__"
    Type "State" is incompatible with type "State | None"
      "scenario.state.State" is incompatible with "_private.harness.State"
      "State" is incompatible with "None" (reportArgumentType)
```

I think maybe this is because in `_private/harness` I import `State` from `ops.testing` if possible and if not (to handle people only installing `ops` and not `ops[testing]`) then I define a placeholder `State` class. This is not pretty but I wasn't sure how else I could have it not complain if ops-scenario wasn't installed (other than an ignore or an `Any`). I think maybe the error now is because it's still using that placeholder class, even though it should be able to get `State` from `ops.testing` (getting it from `scenario` doesn't help).

I'll leave it for someone else (including potentially future-me) to figure out.

Reply:
Oh my bad, this is what I get for looking at code on github instead of in vscode. Thanks for the explanation!